gfx/video_crt_switch.c

```diff
@@ -0,0 +1,218 @@
+		ben_core_height = height;
+		ben_core_hz = hz;
+		check_first_run();
+		
+		if (ben_tmp_height != ben_core_height || ben_core_width != ben_tmp_width){ // detect resolution change and switch
+			screen_setup_aspect(width,height);
+		}
+		
+		if (video_driver_get_aspect_ratio() != fly_aspect){	//check aspect is correct else change
+				video_driver_set_aspect_ratio_value((float)fly_aspect);
+				ben_poke_video();
+		}
+				
+
+}
+
+void check_first_run(){		// ruin of first boot to get current display resolution  
+	if (first_run != 1){
+		orig_height = GetSystemMetrics(SM_CYSCREEN);
+		orig_width = GetSystemMetrics(SM_CXSCREEN);
+	}
+	first_run = 1;
+}
+
+void screen_setup_aspect(int width, int height){ // create correct aspect to fit video if resolution does not exist
+	switch_crt_hz();
+	/* get original resolution of core */	
+	if (width >= 1900){
+		if (height == 4){						//detect menu only
+			height = 480;
+			aspect_ratio_switch(width,height);
+		}
+		if (height < 191 && height != 144){				// for low res games to be dran at res but on a 200 eight
+			aspect_ratio_switch(width,height);
+			height = 200;
+		}	
+		if (height > 191){
+			aspect_ratio_switch(width,height);
+		}
+		if (height == 144 && set_ben_core_hz == 50){				/// for GBA doubling
+			height = 288;
+			aspect_ratio_switch(width,height);
+		}
+		if (height > 200 && height < 224){				/// for mame
+			aspect_ratio_switch(width,height);
+			ben_core_height = 224;
+		}
+		if (height > 224 && height < 240){				/// for mame
+			aspect_ratio_switch(width,height);
+			ben_core_height = 240;
+		}
+	
+		if (height > 240 && height < 255 ){								/// for mame
+			aspect_ratio_switch(width,height);
+			ben_core_height = 254;
+		}
+		if (height == 528 && set_ben_core_hz == 60){								/// for mame
+			aspect_ratio_switch(width,height);
+			ben_core_height = 480;
+		}
+	
+		if (height >= 240 && height < 255 && set_ben_core_hz == 55){
+			aspect_ratio_switch(width,height);
+			ben_core_height = 254;
+		}
+		
+	}
+	switch_res_crt(width, height);
+}
+
+void switch_res_crt(int width, int height){
+		
+	if ( height > 100){
+
+			switch_res(width, height,0);
+			ben_poke_video();
+			ben_tmp_height = ben_core_height;
+			ben_tmp_width = ben_core_width;
+	
+	}
+	
+}
+
+void aspect_ratio_switch(int width,int height){  // send aspect float to videeo_driver
+		fly_aspect = (float)width/height;
+		video_driver_set_aspect_ratio_value((float)fly_aspect);
+}
+
+void switch_crt_hz(){  // set hz float an int for windows switching
+	
+	if (ben_core_hz != ben_tmp_core_hz){
+		if (ben_core_hz < 53 ){
+			set_ben_core_hz = 50;	
+			
+		}
+		if (ben_core_hz >= 53  &&  ben_core_hz < 57){	
+			set_ben_core_hz = 55;	
+		}	
+		if (ben_core_hz >= 57 ){	
+			set_ben_core_hz = 60;	
+		}	
+		video_monitor_set_refresh_rate(ben_core_hz);
+		ben_tmp_core_hz = ben_core_hz;
+	}
+	
+}
+
+void switch_res(int width, int height, int f_restore){  // windows function to swith resolutions
+	
+	DEVMODE curDevmode;

```

Comment:
All of this is Windows-specific and probably Windows-version-specific (e.g. Win95/NT may not support all kinds of live mode switching), regardless it should not be in a generic file like this and broken out to OS-specific backends.